### PR TITLE
Laboratory work #2, Dasha Oleinik - 23FPL1 

### PR DESCRIPTION
**Pull request by dkole13**

### CODE REVIEW

File: lab_2_retrieval_w_bm25/main.py [hunk 1]
```diff
@@ -33,6 +41,14 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
...
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(tokens, list) or not all(isinstance(token, str) for token in tokens) or \
+            not tokens:
+        return None
+    if not isinstance(stopwords, list) or not all(isinstance(word, str) for word in stopwords) or \
+            not stopwords:
+        return None
+
+    return [token for token in tokens if token not in stopwords]
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:51]: nice! 🏅

**Review by EIzharikov — COMMENTED** [review]
Nice start! Fix my one comment and go to 10!!! I wish you good luck!

File: lab_2_retrieval_w_bm25/main.py [hunk 2]
```diff
@@ -47,6 +64,18 @@ def build_vocabulary(documents: list[list[str]]) -> list[str] | None:
...
+            not all(isinstance(document, list) for document in documents) or \
+            not documents:
+        return None
+    for document in documents:
+        if not all(isinstance(item, str) for item in document):
+            return None
+
+    result = set()
+    for doc in documents:
+        result |= set(doc)
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:77]: nice!

File: lab_2_retrieval_w_bm25/main.py [hunk 3]
```diff
@@ -77,6 +117,22 @@ def calculate_idf(vocab: list[str], documents: list[list[str]]) -> dict[str, flo
...
+        return None
+
+    total_documents = len(documents)
+    idf = {}
+    for word in vocab:
+        doc_has_word_count = 0
+        for document in documents:
+            if word in document:
+                doc_has_word_count += 1
+        idf[word] = log((total_documents - doc_has_word_count + 0.5) / (doc_has_word_count + 0.5))
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:134]: why +0.5?

**Review by EIzharikov — COMMENTED** [review]
Nice progress! But you have red CI, fix it. Good luck!

File: lab_2_retrieval_w_bm25/main.py (hunk 3 above)
Comment (dkole13) [reply lab_2_retrieval_w_bm25/main.py:134]: Снимок экрана 2024-10-15 161258
bc the formula has 0.5

**Comment by EIzharikov** [comment]
Everything is great! Be ready to show this lab in next Friday. Good luck!

**Pull request merged**